extensions/amp-form/0.1/amp-form.js

```diff
@@ -436,27 +447,67 @@ export class AmpForm {
    */
   handleXhrSubmit_(varSubsFields, trust) {
     this.setState_(FormState_.SUBMITTING);
+    const varSubPromise = this.doVarSubs_(varSubsFields);
+    if (this.templateRendererHelper_.canRenderTemplates()) {
+      varSubPromise.then(() => {
+        this.actions_.trigger(
+            this.form_, 'submit', /* event */ null, trust);
+      }).then(() => {
+        return this.templateRendererHelper_.fetchAndRenderTemplate(
+            this.form_,
+            this.handleProxyRenderTemplateSuccess_,
+            this.handleProxyRenderTemplateFailure_
+        );
+      });
+    } else {
+      const p = varSubPromise
+          .then(() => {
+            this.handleFormSubmitting_(trust);
+          })
+          .then(() => this.doActionXhr_())
+          .then(response => this.handleXhrSubmitSuccess_(response),
+              error => {
+                return this.handleXhrSubmitFailure_(/** @type {!Error} */(error));
+              });
+      if (getMode().test) {
+        this.xhrSubmitPromise_ = p;
+      }
+    }
+  }
 
-    const p = this.doVarSubs_(varSubsFields)
-        .then(() => {
-          this.triggerFormSubmitInAnalytics_('amp-form-submit');
-          this.actions_.trigger(
-              this.form_, 'submit', /* event */ null, trust);
-          // After variable substitution
-          const values = this.getFormAsObject_();
-          // At the form submitting state, we want to display any template
-          // messages with the submitting attribute.
-          this.renderTemplate_(values);
-        })
-        .then(() => this.doActionXhr_())
-        .then(response => this.handleXhrSubmitSuccess_(response),
-            error => {
-              return this.handleXhrSubmitFailure_(/** @type {!Error} */(error));
-            });
+  /**
+   * Handles viewer render template success.
+   * @param {!JsonObject} response
+   */
+  handleProxyRenderTemplateSuccess_(response) {
```

Comment:
Let's refer to this feature consistently as "SSR" rather than "proxy".

`handleSsrTemplateSuccess_()`